### PR TITLE
Arc-Length Positional Encoding: 1D surface topology for pressure smoothness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current Wavelet-Fourier PE (a major merged breakthrough) applies to all nodes using their 2D (x, y) coordinates. However, surface nodes have a natural **1D ordering**: arc-length from the leading edge going clockwise around the airfoil. The 2D PE treats surface nodes identically to volume nodes, missing this topological structure.

A **surface-specific arc-length PE** explicitly encodes this 1D ordering, helping the model learn smooth Cp distributions along the surface. This is analogous to how 1D sinusoidal PE in language transformers enables smooth sequence understanding — but for the airfoil surface contour.

The arc-length PE is **orthogonal** to the existing 2D spatial PE (they capture different geometric information) and is added as additional features without replacing anything. It's especially useful for:
- Reducing leading/trailing edge pressure oscillations (the 1D ordering constrains smoothness)
- Helping the SRF decoder understand where on the surface each node sits (suction peak vs. trailing edge vs. stagnation)
- For tandem configurations, distinguishing equivalent arc-length positions on fore vs. aft foil

**Key insight:** The surface is a 1D manifold embedded in 2D space. The current PE encodes the 2D embedding but not the 1D intrinsic structure.

## Instructions

**Changes to `cfd_tandemfoil/train.py`:**

Add flags: `--arc_length_pe` and `--arc_pe_dim 16`

### 1. Compute arc-length for surface nodes

In the input feature construction (where surface nodes are identified), for each sample:

```python
# For each foil (fore and aft separately):
# 1. Get surface node indices for this foil
# 2. Find the leading edge node (minimum x-coordinate, or more robustly: 
#    the node closest to the theoretical stagnation point using 
#    score = x * cos(AoA) + y * sin(AoA), take argmin)
# 3. Order surface nodes clockwise from the leading edge using angle from centroid:
#    centroid = mean(x_surf, y_surf)
#    angle = atan2(y - cy, x - cx)
#    Sort by angle starting from the LE node
# 4. Compute cumulative arc-length: 
#    ds = sqrt((x[i+1]-x[i])^2 + (y[i+1]-y[i])^2)
#    s = cumsum(ds) 
#    s = s / s[-1]  # normalize to [0, 1]
```

### 2. Fourier encoding of arc-length

```python
# For each surface node with normalized arc-length s in [0, 1]:
arc_pe = []
for k in range(1, arc_pe_dim // 2 + 1):  # k = 1..8 for dim=16
    arc_pe.append(torch.sin(2 * math.pi * k * s))
    arc_pe.append(torch.cos(2 * math.pi * k * s))
arc_pe = torch.stack(arc_pe, dim=-1)  # [N_surf, arc_pe_dim]
```

### 3. Integrate into the input features

```python
# Create a feature tensor of shape [N_total, arc_pe_dim]
# For surface nodes: fill with the Fourier arc-length features
# For volume nodes: zero-pad (they have no arc-length)
arc_features = torch.zeros(N_total, arc_pe_dim, device=device)
arc_features[surface_node_mask] = arc_pe

# Concatenate to the input features BEFORE the input projection
# This means the input projection dimension increases by arc_pe_dim
x_input = torch.cat([x_input, arc_features], dim=-1)
```

### 4. Handle tandem vs. single-foil

- For **single-foil** samples: compute arc-length for the single foil normally.
- For **tandem** samples: compute arc-length **separately** for fore and aft foil. Each foil has its own [0, 1] normalized arc-length. To help the model distinguish them, add an additional binary indicator: `is_aft_foil` (0 for fore, 1 for aft) as one more feature in the arc-length PE. This makes the total arc_pe_dim = 16 + 1 = 17 features for the arc PE block.

### 5. Flag wiring

```python
parser.add_argument('--arc_length_pe', action='store_true')
parser.add_argument('--arc_pe_dim', type=int, default=16, 
                    help='Fourier feature dimension for arc-length PE')
```

Adjust the input projection `nn.Linear(input_dim + arc_pe_dim + 1, n_hidden)` when `arc_length_pe` is enabled (+17 input channels: 16 Fourier + 1 foil indicator).

### 6. Important details

- Arc-length computation should be done at **data loading time** or at the start of the forward pass (it's a geometric computation, not learned).
- The Fourier frequencies k=1..8 span from 1 full cycle to 8 full cycles around the airfoil. This captures both global (suction side vs. pressure side) and local (LE peak, TE region) patterns.
- Do NOT replace the existing Wavelet-Fourier PE — this is additive. Both PEs carry different information.
- The arc-length PE has **zero learned parameters** (pure deterministic feature) beyond the existing input projection weight matrix size change.
- Make sure the clockwise ordering is consistent across all samples. A simple approach: always start from the LE node and go in the direction of increasing y first (upper surface), then decreasing y (lower surface).

### 7. Run command

```bash
cd cfd_tandemfoil && python train.py \
    --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first \
    --pressure_deep --residual_prediction --surface_refine --te_coord_frame \
    --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
    --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
    --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 \
    --vortex_panel_n 64 \
    --arc_length_pe --arc_pe_dim 16 \
    --wandb_group arc-length-pe
```

Run 2 seeds (42, 73) as standard.

## Baseline

Current baseline (PR #2357, 2-seed avg):
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| p_in | 11.872 | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| p_tan | 26.319 | < 26.319 |
| p_re | 6.229 | < 6.229 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2389 because CLOSING: ANP merged (PR #2379) and the new baseline is dramatically lower. Arc-Length PE was designed for surface topology encoding but ANP's cross-attention already captures surface structure via cross-foil attention. This experiment hasn't started training yet — reassigning thorfinn to a new P0 idea (DeltaPhi Inviscid-to-Viscous Residual) on the ANP baseline..

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Arc-Length PE (dim=16, n_freq=8) — Final Results (2-seed average)

| Metric | Seed 42 | Seed 73 | **2-seed avg** | Baseline | Δ |
|--------|---------|---------|---------------|----------|---|
| **p_in** | 12.9 | 12.4 | **12.65** | 11.872 | **+6.6%** |
| p_oodc | 7.8 | 7.3 | **7.55** | 7.459 | **+1.2%** |
| **p_tan** | 27.1 | 27.0 | **27.05** | 26.319 | **+2.8%** |
| **p_re** | 6.4 | 6.1 | **6.25** | 6.229 | **+0.3%** |

All metrics worse than baseline. Arc-Length PE **failed** — no improvement on any metric.

### Training Details

| Run | Seed | Best epoch | val/loss | W&B ID |
|-----|------|-----------|----------|--------|
| arc-pe-s42 | 42 | 151 | — | iugdyjpc |
| arc-pe-s73 | 73 | 150 | — | r5yydyuq |

**Command used:**
```bash
cd cfd_tandemfoil && python train.py \
  --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --arc_length_pe --arc_pe_dim 16 \
  --wandb_group arc-length-pe \
  --agent thorfinn --wandb_name "thorfinn/arc-pe-s42" \
  --seed 42
```

**Peak memory:** 41.0GB (vs baseline ~40GB — negligible increase from 17 extra input channels)

### What happened

Arc-length PE failed modestly across all metrics. Training was stable — no divergence — but the added features did not improve performance. Analysis:

1. **The 1D topology is already captured.** The existing 2D Fourier PE (wavelet + learned frequencies) with 32 channels already encodes spatial position well enough that the model can infer surface ordering from (x,y). The explicit 1D arc-length is redundant.

2. **17 extra input channels dilute useful signal.** For volume nodes (the majority), these are all zeros — effectively adding noise to the input projection weights. The input projection must accommodate these extra dimensions, and the zeros waste representational capacity.

3. **The SRF already handles surface refinement.** The SurfaceRefinementHead has access to hidden features that encode surface-specific information. Adding arc-length to the backbone input means the information must survive through all Transolver blocks before reaching the SRF — a more targeted approach would add it directly to SRF input.

4. **p_re stayed nearly flat** (+0.3%, within 2-seed noise) while p_in regressed most (+6.6%). This suggests the arc-length features confuse the model on in-distribution predictions more than they help OOD Reynolds generalization.

### Suggested follow-ups

- Add arc-length features only to the SRF input (not the backbone) — where surface-specific information is most directly useful
- Try a much smaller arc_pe_dim (e.g., 4) to reduce dilution
- Consider arc-length PE only for tandem samples (zero for single-foil) — similar to cp_panel_tandem_only
